src/main/java/org/elasticsearch/plugins/repository/LocalPluginRepository.java

```diff
@@ -0,0 +1,433 @@
+            @Override
+            public boolean accept(PluginDescriptor plugin) {
+                // Accepts all plugins
+                return true;
+            }
+        });
+    }
+
+    @Override
+    public Collection<PluginDescriptor> list(final Filter filter) {
+        // Get the es-plugin.properties filename
+        String esPluginProps = PluginsService.ES_PLUGIN_PROPERTIES;
+        if (settings != null) {
+            esPluginProps = settings.get(PluginsService.ES_PLUGIN_PROPERTIES_FILE_KEY, PluginsService.ES_PLUGIN_PROPERTIES);
+        }
+
+        final ImmutableList.Builder<PluginDescriptor> plugins = ImmutableList.builder();
+
+        String pluginName;
+        PluginDescriptor.Builder descriptor;
+
+        if (Files.exists(pluginsDir)) {
+            try (DirectoryStream<Path> root = Files.newDirectoryStream(pluginsDir)) {
+                for (Path pluginDir : root) {
+                    if (Files.isDirectory(pluginDir)) {
+
+                        // Determine the name of the plugin from the directory's name
+                        pluginName = pluginDir.getFileName().toString();
+                        descriptor = pluginDescriptor(pluginName);
+
+                        // Introspect JAR files to find plugin's description
+                        try (DirectoryStream<Path> jarFiles = Files.newDirectoryStream(pluginDir, "*" + JAR)) {
+                            for (Path jarFile : jarFiles) {
+                                // Indicates that the plugin has code/library
+                                descriptor.jvm();
+
+                                String fileName = jarFile.getFileName().toString();
+                                if (!fileName.contains(pluginName)) {
+                                    continue;
+                                }
+
+                                // Constructs the plugin's descriptor from the JAR filename...
+                                // Since plugin's JAR file can have various filenames,
+                                // the result is incertain
+
+                                // Removes "elasticsearch-" prefix
+                                String s = StringUtils.removeStart(fileName, "elasticsearch-");
+
+                                // Removes ".jar" suffix
+                                s = StringUtils.removeEnd(s, JAR);
+
+                                // Removes plugin's name (detected in preVisit() method)
+                                s = StringUtils.removeStart(s, pluginName);
+
+                                // And finally try to extract the version
+                                String version = StringUtils.substringAfterLast(s, "-");
+
+                                if (Strings.hasText(version)) {
+                                    descriptor.version(version);
+                                }
+
+                                // Now, overide everthing with es-plugin.properties or meta.yml file
```

Comment:
I think we should try first `meta.yml` which should content anything (so the classname) and then fallback to old property file.